src/environmentVariables.ts

```diff
@@ -0,0 +1,147 @@
+export type EnvironmentVariablesUndefined = Record<string, string | undefined>; // alias of NodeJS.ProcessEnv
+export type EnvironmentVariables = Record<string, string | undefined | null>;
+export type EnvironmentVariablesItem = EnvironmentVariables | null | undefined;
+
+class EnvironmentVariablesPrivate {
+    private keyMapping: Map<string, string>;
+    public [envProperty]: EnvironmentVariables;
+    protected isWin32: boolean;
+    protected preserveNull: boolean;
+    constructor(
+        preserveNull: boolean | undefined,
+        isWin32: boolean | undefined) {
+        this.keyMapping = new Map<string, string>();
+        this[envProperty] = {};
+        if (preserveNull === undefined) {
+            this.preserveNull = false;
+        } else {
+            this.preserveNull = preserveNull;
+        }
+        if (isWin32 === undefined) {
+            this.isWin32 = os.platform() === 'win32';
+        } else {
+            this.isWin32 = isWin32;
+        }
+    }
+
+    public getKey(key: string, updateKey: boolean): string {
+        if (this.isWin32) {
+            const normalizedKey = key.toUpperCase();
+            let resultKey = this.keyMapping.get(normalizedKey);
+            if (resultKey === undefined) {
+                resultKey = key;
+                if (updateKey) {
+                    this.keyMapping.set(normalizedKey, resultKey);
+                }
+            }
+            return resultKey;
+        }
+        return key;
+    }
+
+    public get(key: string | symbol): string | undefined | null {
+        if (typeof key === 'string') {
+            return this[envProperty][this.getKey(key, false)];
+        }
+        return undefined;
+    }
+
+    public set(key: string | symbol, value?: string | null, receiver?: any): boolean {
+        if (typeof key === 'string') {
+            let needSet = true;
+            if (value === undefined) {
+                needSet = false;
+            } else if (value === null) {
+                if (!this.preserveNull) {
+                    needSet = false;
+                }
+            } else if (typeof value !== 'string') {
+                value = '' + value;
+            }
+            if (needSet) {
+                const existKey = this.getKey(key, true);
+                return Reflect.set(this[envProperty], existKey, value, receiver);
+            } else {
+                return Reflect.deleteProperty(this[envProperty], key);
+            }
+        }
+        return false;
+    }
+
+    [inspect]() {
+        return util.inspect(this[envProperty]);
+    }
+
+    toString() {
+        return this[envProperty].toString();
+    }
+}
+
+export class EnvironmentVariablesUtils {
+
+    public static create(from?: Map<string, string> | EnvironmentVariables | null, preserveNull?: boolean, isWin32?: boolean): EnvironmentVariablesUndefined {
+        const env = new EnvironmentVariablesPrivate(preserveNull, isWin32);
+        const p = new Proxy(env, {
```

Comment:
A lot of complication in this class is because of this Proxy. Can you help me understand why we need to use a Proxy?  I was assuming it was to allow for indexing syntax, but that seems to be [supported in TypeScript](https://www.typescriptlang.org/docs/handbook/2/classes.html#index-signatures) already.

Reply:
> 
EnvironmentVariablesPrivate is proxied because we need
mantain compatiable with NodeJS.ProcessEnv.
For example, supporse we have a env named with `get`, if we using
typescript `Index Signatures`, then what's the result of env.get will have
two meaning:
  * call the function `get`
  * get the environment variable `get`
But for environment variable, access any member with `name` should return the expeted
environment variable for that  `name`